packages/SwingSet/src/vats/timer/vat-timer.js

```diff
@@ -1,98 +1,830 @@
+    }
+  }
+}
+
+/**
+ * @param {Schedule} schedule
+ * @returns {TimestampValue | undefined}
+ */
+function firstWakeup(schedule) {
+  const iter = schedule.keys()[Symbol.iterator]();
+  const first = iter.next();
+  if (first.done) {
+    return undefined;
+  }
+  return first.value;
+}
+
+// if you really must replace "time <= upto" below, use this
+// function timeLTE(a, b) {
+//    return TimeMath.compareAbs(a, b) !== 1;
+// }
+// if (timeLTE(time, upto)) {
+
+/**
+ * return list of events for time <= upto
+ *
+ * @param {Schedule} schedule
+ * @param {TimestampValue} upto
+ * @returns { Event[] }
+ */
+function removeEventsUpTo(schedule, upto) {
+  assert.typeof(upto, 'bigint');
+  let ready = [];
+  for (const [time, events] of schedule.entries()) {
+    if (time <= upto) {
+      ready = ready.concat(events);
+      schedule.delete(time);
+    } else {
+      break; // don't walk the entire future
+    }
+  }
+  return ready;
+}
+
+/*
+ * @param {TimestampValue} start
+ * @param {RelativeTimeValue} interval
+ * @param {TimestampValue} now
+ * @returns {TimestampValue}
+ */
+function nextScheduleTime(start, interval, now) {
+  // used to schedule repeaters
+  assert.typeof(start, 'bigint');
+  assert.typeof(interval, 'bigint');
+  assert.typeof(now, 'bigint');
+  // return the smallest value of `start + N * interval` after now
+  if (now < start) {
+    return start;
+  }
+  return now + interval - ((now - start) % interval);
+}
+
+export function buildRootObject(vatPowers, _vatParameters, baggage) {
   const { D } = vatPowers;
-  const repeaters = new Map();
 
-  async function createTimerService(timerNode) {
-    /** @type {TimerService} */
-    const timerService = Far('timerService', {
-      getCurrentTimestamp() {
-        return Nat(D(timerNode).getLastPolled());
-      },
-      setWakeup(baseTime, handler) {
-        baseTime = TimeMath.toAbs(baseTime);
-        assert(passStyleOf(handler) === 'remotable', 'bad setWakeup() handler');
-        return D(timerNode).setWakeup(baseTime, handler);
-      },
-      // can be used after setWakeup(h) or schedule(h)
-      removeWakeup(handler) {
+  let timerDevice;
+  function insistDevice() {
+    assert(timerDevice, 'TimerService used before createTimerService()');
+  }
+
+  // First, create the KindHandles for all our durable objects. These
+  // are ongoing obligations of the vat: all future versions must
+  // define behaviors for these. We are also obligated to support the
+  // vivifySingleton calls at the bottom: wakeupHandler, timerService,
+  // timerClock, and timerBrand.
+  const oneShotEventHandle = provideKindHandle(baggage, 'oneShotEventHandle');
+  const promiseEventHandle = provideKindHandle(baggage, 'promiseEventHandle');
+  const repeaterEventHandle = provideKindHandle(baggage, 'repeaterEventHandle');
+  const repeaterHandle = provideKindHandle(baggage, 'repeaterHandle');
+  const notifierHandle = provideKindHandle(baggage, 'notifierHandle');
```

Comment:
For every case where we have a call to `provideKindHandle` and a corresponding call to `defineDurableKind` or `defineDurableKindMulti`, I think the code would be clearer to replace these with a single call to `vivifyKind` or `vivifyKindMulti`. Typically, programs will have no other uses of kind handles, e.g., no direct calls to `makeKindHandle`. By making this change, those programs hide the entire concept of a kindHandle, making the durability API easier to learn, and the code using it easier to understand.

For me this was one of the punchlines of seeing our own experimental use of the durability/upgrade API change over time. We gradually climbed abstractions until we got to `provide*` and `vivify*`, therefore initially thinking of them as helpers for saying the other things more compactly. But we now hope most upgradable code can rely exclusively on `provide*` and `vivify*`, typically hiding the need to think about the concepts they encapsulate.